### DISCUSSION TITLE
modeling a multi phase multi-shape initial condition

### DISCUSSION THREAD

**Discussion by Ali-toghraee** [idaholab/moose#16771] [Q&A Modules: Phase field] [answered]
Dear MOOSE users,

I am trying to model a multi phase system (picture attached) that each phase has its own phase and concentration initial conditions. But I have a problem to model this IC. I could not use "SmoothSuperellipsoidIC" or the Circle one because I had already used "SubdomainBoundingBox" mesh modifier and it gave me the error of 
_***** ERROR ***
The initial condition 'c_fe_Oxide' is being defined on a block that already has an initial condition defined.**_
 
Can any body help me please on     how can I implement this IC?

Best,
Alireza
![4phaseIC](https://user-images.githubusercontent.com/77089229/105276336-61280c00-5b5e-11eb-99bd-b930864d996b.png)

**Comment by WilkAndy** [comment, marked as answer]
Hi Alireza,

Could you quickly run your input file with the `--mesh-only` flag, load the result into paraview, and check that your subdomains (blocks) are actually as you desire?

a

**Reply by Ali-toghraee** [reply]
Hi Andy,

I did that and it showed the 3 blocks as I had defined them in mesh modifier.  But no ellipsoid in the middle  

A

**Reply by WilkAndy** [reply]
So, if you need the 4th ellipsoidal block, you're going to have to use something like a `ParsedSubdomainMeshModifier`.  With 4 blocks, you can define an IC for each block separately.

**Reply by Ali-toghraee** [reply]
I am looking in "https://mooseframework.inl.gov/source/" and I could find  "ParsedSubdomainMeshGenerator"  did you mean that? 
If not, can you guide to where I can find "ParsedSubdomainMeshModifier" and some example input files of that?

Thanks,
A

**Reply by WilkAndy** [reply]
Oh, i'm sorry, yes, i meant `ParsedSubdomainMeshGenerator`

**Reply by Ali-toghraee** [reply]
Hello again,

So, I followed your instruction to use "ParsedSubdomainMeshGenerator" , (I have attached the part of code related to that) but I keep getting the error that:
*** ERROR ***
4eta_cirIC.i:2: section 'Mesh/gmc' does not have an associated "Action".
Did you misspell it?
4eta_cirIC.i:16: section 'Mesh/subdomains' does not have an associated "Action".
Did you misspell it?


I have tried different combination of  [ ] [ ] and  [./] [../] for the gmc and the subdomains part but still same error. 
I was wondering if you could help on this.

Thanks,
Ali

[4eta_cirIC.txt](https://github.com/idaholab/moose/files/5853537/4eta_cirIC.txt)

**Reply by WilkAndy** [reply]
It's almost right.  Use `[./gmc]` instead of `[gmc]`.  Use `[./subdomains]` instead of `[subdomains]`.  Also, remove the `MeshModifiers` and put them all in the `[Mesh]` block using a number of `SubdomainBoundingBoxGenerator`.

**Reply by Ali-toghraee** [reply]
I changed my input to:

[Mesh]
  [./gmc]
    type =  GeneratedMeshGenerator
    dim = 2
    nx = 100
    ny = 100
    xmax = 1000
    xmin = 0
    ymax = 1000
    ymin = 0
    elem_type = QUAD4
  []

   [./subdomaincir]
    type = ParsedSubdomainMeshGenerator
    input = gmc
    combinatorial_geometry = '(x-400)^2+(y-500)^2<25'
    block_id = 4
   []

    [./Fe_bot]
      type = SubdomainBoundingBoxGenerator
      block_id = 1
      block_name = bot
      bottom_left = '0 0 0'
      top_right = '400 500 0'
    [../]

    [./Fe_top]
      type = SubdomainBoundingBoxGenerator
      block_id = 2
      block_name = up
      bottom_left = '0 500 0'
      top_right = '400 1000 0'
    [../]

    [./air]
      type = SubdomainBoundingBoxGenerator
      block_id = 3
      block_name = air
      bottom_left = '400 0 0 '
      top_right = '1000 1000 0'
    [../]

[]

But I still get the same error!

*** ERROR ***
4eta_cirIC.i:2: section 'Mesh/gmc' does not have an associated "Action".
Did you misspell it?
4eta_cirIC.i:14: section 'Mesh/subdomaincir' does not have an associated "Action".
Did you misspell it?
4eta_cirIC.i:21: section 'Mesh/Fe_bot' does not have an associated "Action".
Did you misspell it?
4eta_cirIC.i:29: section 'Mesh/Fe_top' does not have an associated "Action".
Did you misspell it?
4eta_cirIC.i:37: section 'Mesh/air' does not have an associated "Action".
Did you misspell it?

**Reply by WilkAndy** [reply]
Hey,

Hopefully this helps...  This is my input file:

```
[Mesh]
[./gmc]
type = GeneratedMeshGenerator
dim = 2
nx = 100
ny = 100
xmax = 1000
xmin = 0
ymax = 1000
ymin = 0
elem_type = QUAD4
[]

[./Fe_bot]
  type = SubdomainBoundingBoxGenerator
  input = gmc
  block_id = 1
  block_name = bot
  bottom_left = '0 0 0'
  top_right = '400 500 0'
[../]

[./Fe_top]
  type = SubdomainBoundingBoxGenerator
  input = Fe_bot
  block_id = 2
  block_name = up
  bottom_left = '0 500 0'
  top_right = '400 1000 0'
[../]

[./air]
  type = SubdomainBoundingBoxGenerator
  input = Fe_top
  block_id = 3
  block_name = air
  bottom_left = '400 0 0 '
  top_right = '1000 1000 0'
[../]
[./subdomaincir]
type = ParsedSubdomainMeshGenerator
input = air
  combinatorial_geometry = '((x-400)/0.5)^2+(y-500)^2<10000'
block_id = 4
[]


[]
```

I run it with this command:

```
myapp-opt -i tmp.i --mesh-only
```

and it produces this figure:

![tmp](https://user-images.githubusercontent.com/6676821/105450117-8c793b00-5cc5-11eb-9c32-2dda5a3b474c.png)

**Reply by Ali-toghraee** [reply]
Thanks Andy,

I copy and pasted your input but it still shows the same error. Is this a new feature added to moose? since I have not updated the moose in last 12 months.

Best
Ali

**Reply by WilkAndy** [reply]
Yes, that is almost definitely the problem.   You can continue with the old MOOSE but the above stuff won't work as i'm pretty sure the MeshGenerator stuff wasn't completed at that time.

**Reply by Ali-toghraee** [reply]
Hello again Andy,

Thanks for your helps so far. I have another question.
I am trying to create a semi poly-crystalline domain. Some thing like the pic below. 

![mesh](https://user-images.githubusercontent.com/77089229/109598063-bc420c80-7ad5-11eb-8ea9-7b719c494d86.png)

where each color is a separate block (phase). 
-Do you think that there is a way to do this using my old version of MOOSE :) ? 
(It is near my PhD defense and I haven't had very good experience updating MOOSE as it takes time for me to get use to all of the changes) 
- If updating is the only way, is it possible that I save my current version of MOOSE some how? 

Thanks again!
Ali

**Reply by aeslaughter** [reply]
It is possible to create the subdomains from an image: https://mooseframework.inl.gov/source/meshgenerators/ImageSubdomainGenerator.html

**Reply by aeslaughter** [reply]
>     * If updating is the only way, is it possible that I save my current version of MOOSE some how?

Yes. You can always checkout a version of MOOSE given the commit hash (use "git log") to get that. You can also create a branch from you working commit: "git branch some-usefull-name-here". Then if you create a [fork of MOOSE](https://docs.github.com/en/github/getting-started-with-github/fork-a-repo) you can push that branch to your fork and then it will be stored for your future access, even if you lose you computer.

**Reply by Ali-toghraee** [reply]
Thank you aeslaughter,
I checked that, but it seems since my MOOSE is old I can not run this 

bellow is my input file:

[Mesh]
  [gen]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 200
    ny = 100
  []
  [./image]
    input = gen
    type = ImageSubdomainGenerator
    file = mesh00.png
    threshold = 100
  [../]
[]

[Variables]
  [u]
  []
[]

[Problem]
  type = FEProblem
  solve = false
[]

[Executioner]
  type = Steady
[]

[Outputs]
  execute_on = 'timestep_end'
  exodus = true
[]

 And I run it using the --mesh-only  flag in command line but I get the following error.


atoghraee@ME-DESKTOP102:~/projects/Oxidation_poly/metal_bi_t1180/mesh$ mpiexec -np 88 /home/atoghraee/projects/moose/modules/combined/combined-opt -i mesh.i --mesh-only


 *** ERROR ***
mesh.i:4: section 'Mesh/gen' does not have an associated "Action".
Did you misspell it?
mesh.i:10: section 'Mesh/image' does not have an associated "Action".
Did you misspell it?

**Reply by Ali-toghraee** [reply]
About saving my current version of MOOSE, could you please explain that a little bit less "githuby" :) . 
Sorry to ask that, but I am more mechanical and dont have much experience with Linux and github and their expressions.
 I dont have any repository on github (I think it is different than having an account) and all of my stuff are on my computer.
So far I:
- opened a terminal on my moose folder ( the one I downloaded and installed on Feb 2019) and then typed "git log" which gave me a  lot of log information. 
- then I did the "git branch  some_good_name.  Should it create anything on my computer? because I dont see any file or folder created under the given name. 
-  I am reading the link you sent about how to do a "Fork a repo"  and will be back to ask questions. I hope I am on the right track.

Thanks again for the helps and sorry if I ask elementary questions.

Ali

**Reply by aeslaughter** [reply]
I think it would be best if you spend a bit of time learning git, which is a version control tool. Github is just a web-based tool for managing a git repository. Learning git will be invaluable beyond just working with MOOSE. There are lots of online tutorials:

https://git-scm.com/book/en/v2
https://guides.github.com/
